### PR TITLE
opt: support use of streaming set operations without required ordering

### LINKED ISSUE #64062: opt: support streaming set ops when ordering not required
#63805 adds support for planning streaming set operations (UNION, INTERSECT, EXCEPT + ALL variants) when an ordering is required by a parent operator or by an ORDER BY in the SQL query. However, we should also be able to plan streaming set operations if the ordering is not required, since it may still produce a more efficient plan. To make this work, as described [here](https://github.com/cockroachdb/cockroach/pull/63805#pullrequestreview-639298356), we need exploration rules that generate unions with interesting orderings, similar to streaming groupby, and the coster needs to make the distinction between the hash and the merge execution.

### PR DESCRIPTION
**Pull request by rytaft**
This commit adds a new rule, `GenerateStreamingSetOp`, which creates variants
of set operations that require more specific orderings on the input columns,
using the interesting orderings property. When we have orderings on the input
columns, we can execute set operations in a streaming fashion which is more
efficient.

Fixes #64062

Release note (performance improvement): The optimizer can now create query
plans that use streaming set operations, even when no ordering is required
by the query. Streaming set operations are more efficient than the alternative
hash set operations, because they avoid the overhead of building a hash table.
This can result in improved performance for queries containing the set
operations UNION, INTERSECT, INTERSECT ALL, EXCEPT, and EXCEPT ALL.
(UNION ALL does not benefit from this optimization.)

### CODE REVIEW

**Review by rytaft — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64953)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @RaduBerinde)

___
*[pkg/sql/opt/xform/testdata/physprops/ordering, line 1542 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-M_Mt-GdBV_LxjBCDdLC:-M_Mt-Ge1cRnaAYt-hr8:b-6zbffh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0d7f11e9083c7aaf991df5f135cafb31c61c2d9/pkg/sql/opt/xform/testdata/physprops/ordering#L1542)):*
> ```
>  ├── left columns: a:1 b:2 c:3 d:4
>  ├── right columns: c:9 d:10 a:7 b:8
>  ├── internal-ordering: +3,+4
> ```

For groupby it seems like we don't show the `internal-ordering` if it's the same as the ordering. Couldn't figure out where that logic is, though -- do you know, @raduberinde?
___
*[pkg/sql/opt/xform/testdata/rules/limit, line 1194 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-M_MtP2j2WruWft_hD3t:-M_MtP2j2WruWft_hD3u:b7m1w1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0d7f11e9083c7aaf991df5f135cafb31c61c2d9/pkg/sql/opt/xform/testdata/rules/limit#L1194)):*
> ```
>  ├── cardinality: [0 - 10]
>  ├── ordering: +6,+7
>  ├── sort
> ```

This seems like it might be a worse plan -- I'll dig into this a bit to understand what's going on

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64953)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/xform/set_funcs.go, line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-M_Mvn-F9UcqRv4cTPYj:-M_Mvn-F9UcqRv4cTPYk:bcvpvk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0d7f11e9083c7aaf991df5f135cafb31c61c2d9/pkg/sql/opt/xform/set_funcs.go#L39)):*
> ```Go
> ) {
> 	orders := ordering.DeriveInterestingOrderings(grp)
> 	for i := range orders {
> ```

Conceptually we either have a non-streaming op, or a streaming op which has an ordering on all columns. The operators should map to these concepts: a streaming set op should always contain a "full" ordering that involves all output columns. (this is a little tricky to define with OrderingChoice but I think something that mentions all columns regardless of how they're split into groups/optional would work).

In addition to making things easier to reason about, this would allow simplifying setOpBuildRequired and simplifying/removing StreamingSetOpOrdering().

**Comment by RaduBerinde** [comment]
___
*[pkg/sql/opt/xform/testdata/physprops/ordering, line 1542 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-M_Mt-GdBV_LxjBCDdLC:-M_MyD0Q0A6766E4LvzB:b-z776ox) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0d7f11e9083c7aaf991df5f135cafb31c61c2d9/pkg/sql/opt/xform/testdata/physprops/ordering#L1542)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

For groupby it seems like we don't show the `internal-ordering` if it's the same as the ordering. Couldn't figure out where that logic is, though -- do you know, @raduberinde?
</blockquote></details>

We always show it (eg https://github.com/cockroachdb/cockroach/blob/52da7bf8923567460d0c5a7dee5b8af3a98e3429/pkg/sql/opt/xform/testdata/physprops/ordering#L1032
)

**Review by rytaft — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-:-M_N4NHV6cd-qvqyAD4k:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/opt/xform/set_funcs.go, line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-M_Mvn-F9UcqRv4cTPYj:-M_N3uV382yZPjd6Z_SY:bcsn8tr) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0d7f11e9083c7aaf991df5f135cafb31c61c2d9/pkg/sql/opt/xform/set_funcs.go#L39)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Conceptually we either have a non-streaming op, or a streaming op which has an ordering on all columns. The operators should map to these concepts: a streaming set op should always contain a "full" ordering that involves all output columns. (this is a little tricky to define with OrderingChoice but I think something that mentions all columns regardless of how they're split into groups/optional would work).

In addition to making things easier to reason about, this would allow simplifying setOpBuildRequired and simplifying/removing StreamingSetOpOrdering().
</blockquote></details>

Done. I wasn't able to simplify `setOpBuildRequired` or `StreamingSetOpOrdering`, though -- removing that logic produces an incorrect plan in some cases.
___
*[pkg/sql/opt/xform/testdata/physprops/ordering, line 1542 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-M_Mt-GdBV_LxjBCDdLC:-M_N-cIt8Jc-YJd9vEFN:by48yjg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0d7f11e9083c7aaf991df5f135cafb31c61c2d9/pkg/sql/opt/xform/testdata/physprops/ordering#L1542)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We always show it (eg https://github.com/cockroachdb/cockroach/blob/52da7bf8923567460d0c5a7dee5b8af3a98e3429/pkg/sql/opt/xform/testdata/physprops/ordering#L1032
)
</blockquote></details>

Oh ok thanks. I guess I was just thrown off by the `internal-ordering` getting removed from some of the tests in this PR.
___
*[pkg/sql/opt/xform/testdata/rules/limit, line 1194 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-M_MtP2j2WruWft_hD3t:-M_MvtCvECq-2akQm7F8:b-bao3oe) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0d7f11e9083c7aaf991df5f135cafb31c61c2d9/pkg/sql/opt/xform/testdata/rules/limit#L1194)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

This seems like it might be a worse plan -- I'll dig into this a bit to understand what's going on
</blockquote></details>

Actually, I think it's pretty similar. The scans that used to have a soft limit all still have a hard limit. I don't think we really do anything with the limit hint on UNION (at least we don't account for it in the coster), so it doesn't seem like a problem that we've removed those hints. We estimate that the constrained scan of `index_tab@d` produces 2000 rows, while the tree of UNIONs produces 2030 rows. Moving the sort above the UNION tree allows us to sort on only two columns, while the previous plan required sorting on four columns. 

TL;DR I'm not going to worry too much about this case.

**Review by RaduBerinde — COMMENTED** [review]
This would be a brilliant time to address the ordering TODO in the execbuild of set ops. To make this more clear at the vital interface between opt and execution, we should create two execution operators - a HashSetOp and a StreamingSetOp. Only the latter would take the "streaming ordering" and the "output required ordering".

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-:-M_N9A--BtLohPKyXbcp:ba8ddu8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/ordering/set.go, line 76 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-M_NBKPa7BeAgEVKtavH:-M_NBKPa7BeAgEVKtavI:bu4bjwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3dc9add524a08b93c77198a9236b2b1b8ce360b1/pkg/sql/opt/ordering/set.go#L76)):*
> ```Go
> 
> 	result := required.Intersection(&private.Ordering)
> ```

We should probably always Simplify here. The child may have "stronger" FDs that don't carry over to the set op.
___
*[pkg/sql/opt/xform/set_funcs.go, line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-M_Mvn-F9UcqRv4cTPYj:-M_N7wBCEFNdwtVvYUfq:bzct0p6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0d7f11e9083c7aaf991df5f135cafb31c61c2d9/pkg/sql/opt/xform/set_funcs.go#L39)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Done. I wasn't able to simplify `setOpBuildRequired` or `StreamingSetOpOrdering`, though -- removing that logic produces an incorrect plan in some cases.
</blockquote></details>

I think the ordering might still not reference all the columns.. If `x` is a key, `Simplify()` would change `+x,+y` to `+x`. I think maybe we should check again after Simplify and add extra groups at the end.

Actually, it would be better to use an `Ordering` instead of an `OrderingChoice` in the operator, to localize this kind of weirdness to a single place. The `Ordering` would be empty or have as many columns as output columns. We definitely wouldn't need `StreamingSetOpOrdering` anymore because that would just be the ordering in the operator.

**Review by RaduBerinde — COMMENTED** [review]
Well, actually three operators:
 - HashSetOp (takes no orderings)
 - StreamingSetOp (takes streaming ordering and req output ordering)
 - UnionAllOp (takes req output ordering)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-:-M_ND1eR2CBntbX4FM-8:bgpygtp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

**Review by mgartner — COMMENTED** [review]
Nice!

> Reviewed 11 of 22 files at r1, 11 of 11 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-:-M_RQhG06JLzDWDfE8KR:b-83sjoh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft)

___
*[pkg/sql/opt/memo/check_expr.go, line 357 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-M_RIiAL4WVQqxqiI2wi:-M_RIiAL4WVQqxqiI2wj:b-5hzctb) ([raw file](https://github.com/cockroachdb/cockroach/blob/3dc9add524a08b93c77198a9236b2b1b8ce360b1/pkg/sql/opt/memo/check_expr.go#L357)):*
> ```Go
> func checkExprOrdering(e opt.Expr) {
> 	// Verify that orderings stored in operators only refer to columns produced by
> 	// their input.
> ```

nit: this comment doesn't match the code. it should say something like "Verify that orderings stored in operators only refer to output columns."
___
*[pkg/sql/opt/xform/set_funcs.go, line 42 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-M_RMQg1BEiFZZTF5whN:-M_RMQg1BEiFZZTF5whO:b9hlejc) ([raw file](https://github.com/cockroachdb/cockroach/blob/3dc9add524a08b93c77198a9236b2b1b8ce360b1/pkg/sql/opt/xform/set_funcs.go#L42)):*
> ```Go
> 		// A streaming set operation must have an ordering that includes all output
> 		// columns. If the interesting ordering includes some columns but not all,
> 		// add the remaining columns in an arbitrary (but deterministic) order.
> ```

I'm confused why it is safe to add columns to an interesting ordering. If `left` or `right` cannot provide such an ordering, then a sort would have to wrap each, but I don't see how that will happen when calling `c.e.mem.Add[SetOp]ToGroup`.

Also, I don't see any test cases that cover this case when `missing` is non-empty.

**Comment by mgartner** [comment]
___
*[pkg/sql/opt/xform/set_funcs.go, line 42 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-M_RMQg1BEiFZZTF5whN:-M_RSeyd3_oiVkmvvPx4:bbs0p2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/3dc9add524a08b93c77198a9236b2b1b8ce360b1/pkg/sql/opt/xform/set_funcs.go#L42)):*
> Also, I don't see any test cases that cover this case when missing is non-empty.

Disregard this - I think your tests do cover this.

I guess I'm generally confused why it is legal for expressions in the same memo group to produce different orderings, especially when the group is a child of another group with an expression that requires a specific ordering.

From one of your tests:

```
memo (optimized, ~9KB, required=[presentation: k:11,u:12,v:13,w:14])
 ├── G1: (union G2 G3) (union G2 G3 ordering=+11,+12,+13,+14) (union G2 G3 ordering=+12,+13,+14,+11) (union G2 G3 ordering=+14,+13,+12,+11) (union G2 G3 ordering=+13,+14,+11,+12) (union G2 G3 ordering=+14,+11,+12,+13)
 │    └── [presentation: k:11,u:12,v:13,w:14]
 │         ├── best: (union G2="[ordering: +1]" G3="[ordering: +6]" ordering=+11,+12,+13,+14)
 │         └── cost: 2209.63
 ├── G2: (scan kuvw,cols=(1-4)) (scan kuvw@uvw,cols=(1-4)) (scan kuvw@wvu,cols=(1-4)) (scan kuvw@vw,cols=(1-4)) (scan kuvw@w,cols=(1-4))
 │    ├── [ordering: +1]
 │    │    ├── best: (scan kuvw,cols=(1-4))
 │    │    └── cost: 1094.81
 │    ├── [ordering: +2,+3,+4,+1]
 │    │    ├── best: (scan kuvw@uvw,cols=(1-4))
 │    │    └── cost: 1094.81
...
```

What happens if the chosen expression in `G1` is `(union G2 G3 ordering=+11,+12,+13,+14)` (`+k,+u,+v,+w`) and the chosen expression in `G2` is `(scan kuvw@uvw,cols=(1-4))` with `[ordering: +2,+3,+4,+1]` (`+u,+v,+w+k`)?

**Review by rytaft — COMMENTED** [review]
Thanks for the reviews, and sorry for the long delay!

> Well, actually three operators

Done.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-:-Mc9qjJQE3dJDWsGI2RO:b-4f1e7t)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @RaduBerinde)

___
*[pkg/sql/opt/memo/check_expr.go, line 357 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-M_RIiAL4WVQqxqiI2wi:-Mc9qcq-3Z5KyrCJNy2Y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3dc9add524a08b93c77198a9236b2b1b8ce360b1/pkg/sql/opt/memo/check_expr.go#L357)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: this comment doesn't match the code. it should say something like "Verify that orderings stored in operators only refer to output columns."
</blockquote></details>

Done.
___
*[pkg/sql/opt/ordering/set.go, line 76 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-M_NBKPa7BeAgEVKtavH:-McAyL_69b0MZ2bKfAv9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3dc9add524a08b93c77198a9236b2b1b8ce360b1/pkg/sql/opt/ordering/set.go#L76)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We should probably always Simplify here. The child may have "stronger" FDs that don't carry over to the set op.
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/set_funcs.go, line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-M_Mvn-F9UcqRv4cTPYj:-McAzUbb4qLEVi7fm0y5:b-6e1gl6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0d7f11e9083c7aaf991df5f135cafb31c61c2d9/pkg/sql/opt/xform/set_funcs.go#L39)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think the ordering might still not reference all the columns.. If `x` is a key, `Simplify()` would change `+x,+y` to `+x`. I think maybe we should check again after Simplify and add extra groups at the end.

Actually, it would be better to use an `Ordering` instead of an `OrderingChoice` in the operator, to localize this kind of weirdness to a single place. The `Ordering` would be empty or have as many columns as output columns. We definitely wouldn't need `StreamingSetOpOrdering` anymore because that would just be the ordering in the operator.
</blockquote></details>

I added a note to explain that the columns might be removed by `Simplify`.

I'm inclined not to change to `Ordering` since there are now a lot of places that can take advantage of the ordering choice without converting back and forth between the two. I started making the change to see what it would look like, but it was making a lot of other parts of the code messier without an obvious benefit. Also seems like it might produce suboptimal plans in some cases.

Although we still need `StreamingSetOpOrdering`, I don't think it's a big deal since it's only called once from the `execbuilder`.
___
*[pkg/sql/opt/xform/set_funcs.go, line 42 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-M_RMQg1BEiFZZTF5whN:-Mc9nNGq-4eSdSn4ry6W:b-i39kj4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3dc9add524a08b93c77198a9236b2b1b8ce360b1/pkg/sql/opt/xform/set_funcs.go#L42)):*
> I guess I'm generally confused why it is legal for expressions in the same memo group to produce different orderings, ...

Expressions in the same memo group don't have to have the same physical properties -- they are only guaranteed to have the same logical properties.

And expressions that require certain physical properties of their children can only select expressions from the child group(s) that have those properties. It's always possible to find such an expression since you can add an enforcer (e.g., sort) to create a new expression in the same group.

This is definitely confusing, so happy to chat more... (also maybe I didn't understand the comment)

**Review by mgartner — APPROVED** [review]
:lgtm: 

> Reviewed 1 of 22 files at r1, 26 of 26 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-:-McGGpQD6_1TydFFKjA-:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @RaduBerinde)

**Comment by rytaft** [comment]
TFTR, @mgartner!

@raduberinde, did you want to take another look?

**Review by RaduBerinde — COMMENTED** [review]
:lgtm_strong: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64953#-:-McPkpXc0m8LbdQ5cb6L:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @RaduBerinde)

**Comment by rytaft** [comment]
TFTRs!

bors r+
